security/pkg/k8s/tokenreview/k8sauthn.go

```diff
@@ -22,34 +22,21 @@ import (
 	k8sauth "k8s.io/api/authentication/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
-
-	"istio.io/istio/pkg/jwt"
-	"istio.io/pkg/env"
-)
-
-var (
-	// TokenAudiences specifies a list of audiences for SDS trustworthy JWT. This is to make sure that the CSR requests
-	// contain the JWTs intended for Citadel.
-	TokenAudiences = strings.Split(env.RegisterStringVar("TOKEN_AUDIENCES", "istio-ca",
-		"A list of comma separated audiences to check in the JWT token before issuing a certificate. "+
-			"The token is accepted if it matches with one of the audiences").Get(), ",")
 )
 
 // ValidateK8sJwt validates a k8s JWT at API server.
 // Return {<namespace>, <serviceaccountname>} in the targetToken when the validation passes.
 // Otherwise, return the error.
 // targetToken: the JWT of the K8s service account to be reviewed
-// jwtPolicy: the policy for validating JWT.
-func ValidateK8sJwt(kubeClient kubernetes.Interface, targetToken, jwtPolicy string) ([]string, error) {
+// aud: list of audiences to check. If empty 1st party tokens will be checked.
+func ValidateK8sJwt(kubeClient kubernetes.Interface, targetToken string, aud []string) ([]string, error) {
 	tokenReview := &k8sauth.TokenReview{
 		Spec: k8sauth.TokenReviewSpec{
 			Token: targetToken,
 		},
 	}
-	if jwtPolicy == jwt.PolicyThirdParty {
-		tokenReview.Spec.Audiences = TokenAudiences
-	} else if jwtPolicy != jwt.PolicyFirstParty {
-		return nil, fmt.Errorf("invalid JWT policy: %v", jwtPolicy)
+	if aud != nil {
+		tokenReview.Spec.Audiences = aud
```

Comment:
From the spec: `If no audiences are provided, the audience will default to the     audience of the Kubernetes apiserver.`

Based on this, my interpretation is for 1p jwt we will set the audience to empty. The 1p jwt token does not have an audience set. The api-server will then assert the aud of the 1p jwt token is equal to the audience of the api-server and fail. Therefor, I expect 1p jwt token to fail.

We have tests for 1p jwt tokens, so I am a bit confused how they are passing. Any idea? Is the api docs wrong/confusing?

Reply:
Actually just ran a request locally:
```
echo '{"kind":"TokenReview","apiVersion":"authentication.k8s.io/v1","spec":{"token":"token"}}' | kubectl create --raw /apis/authentication.k8s.io/v1/tokenreviews -f - | jq
```

Passes for 1p jwt, fails for 3p jwt with `"[invalid bearer token, token audiences [\"istio-ca\"] is invalid for the target audiences [\"kubernetes.default.svc\"]]`

So somehow the 1p jwt gets the `kubernetes.default.svc` aud set even if the token itself doesn't have it?

Reply:
I think token with no audience is equivalent with "kubernentes.default.svc".